Added and implemented delta time

GarrettBassen opened the pull request:
# Description

Major engine changes and reworks to allow for similar play on different devices. Movement and animations are now time-dependent using delta time as opposed to frame-dependent. This allows the game to run the same without need for locking frame rate with former sync() function or vsync. Fixed update was added which is updated 60 times per second regardless of frame rate.


# Fixes and Changes

Timer.java:
Removed unused and redundant functions and variables. Added delta time to make calculations time dependent instead of frame dependent. Made functions static

Player.Java:
Removed redundant import
Changed variable speeds
Changed delta position directions
Added delta time in movement to make player movement time-dependent instead of frame-dependent

Animation.java:
Formatted file
Removed redundant and unused variables
Renamed variables and functions for readability
Fixed run() to use new delta time

GameEngine.java:
Removed redundant and unused variables
Frame unlocked update()
Added fixedUpdate()
Implemented delta time in fixed update
Removed sync() as it was redundant and overridden by glfwSwapInterval(1) Implemented delta time in FPS counter

GameObject.java:
Formatted variables

GameLogic.java:
Frame-unlocked update function

Game.java:
Changed update() to work with new changes.
Made camera follow player

Main.java:
Formatted GameEngine() arguments for readability

MapTree.java:
Small formatting

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

# Possible Bugs and Errors

Animations, while called in fixedUpdate(), are slightly slower when vsync is enabled compared to when frame rate is unlocked. Movement does not appear to be effected by this bug and there does not appear to be any impact on game play.

Tests were conducted to ensure movement speed was the same among multiple frame rates. Further testing on separate devices is encouraged.

GarrettBassen committed:
Added and implemented delta time

Major engine changes and reworks:

Timer.java:
Removed unused and redundant functions and variables.
Added delta time to make calculations time dependent instead of frame dependent.
Made functions static

Player.Java:
Removed redundant import
Changed variable speeds
Changed delta position directions
Added delta time in movement to make player movement time-dependent instead of frame-dependent

Animation.java:
Formatted file
Removed redundant and unused variables
Renamed variables and functions for readability
Fixed run() to use new delta time

GameEngine.java:
Removed redundant and unused variables
Frame unlocked update()
Added fixedUpdate()
Implemented delta time in fixed update
Removed sync() as it was redundant and overridden by glfwSwapInterval(1)
Implemented delta time in FPS counter

GameObject.java:
Formatted variables

GameLogic.java:
Frame-unlocked update function

Game.java:
Changed update() to work with new changes.
Made camera follow player

Main.java:
Formatted GameEngine() arguments for readability

MapTree.java:
Small formatting

JacksonHoggard commented on src/main/java/com/github/jacksonhoggard/voodoo2d/engine/Timer.java:
```diff
@@ -2,24 +2,20 @@
 
     public static double getTime() {
-        return System.nanoTime() / 1000_000_000.0;
+        return System.nanoTime() / 1000000000f; // Time in seconds
```
```suggestion
        return System.nanoTime() / 1000000000D; // Time in seconds
```

JacksonHoggard commented on src/main/java/com/github/jacksonhoggard/voodoo2d/engine/GameEngine.java:
```diff
@@ -44,34 +40,24 @@ public void run() {
 
-            if (!window.isvSync()) {
-                sync();
+            // Fixed update gets called 60 times a second
+            time += Timer.getDeltaTime();
+            if (time > 0.016) {
+                fixedUpdate();
+                time = 0;
```
fixedUpdate should be called before update. This seems to fix the stuttering animations and brings the frame rate up to 60.

JacksonHoggard reviewed (changes requested):
A very welcome change to the engine, the fixedUpdate method will be especially useful later on. Check out the comments I made and make any necessary changes and then I'll merge the changes. Thanks!

GarrettBassen committed:
Fixed update and timer changes

Changed timer to use double instead of float

Put fixed update function call before other calls to fix stuttering and animation issues

GarrettBassen commented:
Changes were made. Happy thanksgiving!

JacksonHoggard reviewed (approved):
Nice work! I'm grateful for all of your contributions recently. Happy Thanksgiving!
